File: two_sum_ds.rb
class TwoSum
    attr_reader :nums

  def initialize
    @nums = {}
  end

  def add(input)
    @nums[input] ? @nums[input] += 1 : @nums[input] = 1
  end

  def find(value)
    @nums.keys.each do |num|
      diff = value - num

      if diff == num
        return true if @nums[num] >= 2
      elsif @nums[diff]
        return true
      end
    end

    return false
  end
end

two_sum = TwoSum.new()
two_sum.add(1)
two_sum.add(3)
two_sum.add(5)
p two_sum.find(4)
p two_sum.find(7)

File: twoSumDS.js
class TwoSumDS {
  constructor() {
    this.nums = {};
  }

  add(input) {
    if (this.nums[input]) {
      this.nums[input] += 1;
    } else {
      this.nums[input] = 1;
    }
  }

  find(value) {
    const keys = Object.keys(this.nums);
    
    for (let i = 0; i < keys.length; i++) {
      let num = keys[i];
      let diff = value - num;
      
      if (diff == num) {
        if (this.nums[num] >= 2) return true;
      } else if (this.nums[diff]) {
        return true;
      }
    }

    return false;
  }
}

let twoSum = new TwoSumDS();
twoSum.add(1);
twoSum.add(3);
twoSum.add(5);
console.log(twoSum.find(4));
console.log(twoSum.find(7));
